File: uploader/src/main/scala/net/hearthstats/ui/DecksTab.scala
package net.hearthstats.ui

import java.awt.BorderLayout
import java.io.IOException
import javax.swing._

import grizzled.slf4j.Logging
import net.hearthstats.Constants._
import net.hearthstats._
import net.hearthstats.util.Translations.t
import net.hearthstats.util.{Browse, HsRobot}
import net.miginfocom.swing.MigLayout
import org.json.simple.JSONObject

import scala.collection.JavaConversions._
import scala.swing.Swing._

class DecksTab(val monitor: Monitor) extends JPanel with Logging {

  val deckSlotComboBoxes = 1 to 9 map { new DeckSlotPanel(_) }

  setLayout(new MigLayout)

  add(new JLabel(" "), "wrap")
  add(new JLabel(t("set_your_deck_slots")), "skip, span 2")
  add(new HelpIcon("https://github.com/HearthStats/HearthStats.net-Uploader/wiki/Decks-Tab", "Help on Decks tab"), "right,wrap")

  add(new JLabel(" "), "wrap")
  add(deckSlotComboBoxes(0), "skip")
  add(deckSlotComboBoxes(1), "")
  add(deckSlotComboBoxes(2), "wrap")

  add(new JLabel(" "), "wrap")
  add(deckSlotComboBoxes(3), "skip")
  add(deckSlotComboBoxes(4), "")
  add(deckSlotComboBoxes(5), "wrap")

  add(new JLabel(" "), "wrap")
  add(deckSlotComboBoxes(6), "skip")
  add(deckSlotComboBoxes(7), "")
  add(deckSlotComboBoxes(8), "wrap")

  add(new JLabel(" "), "wrap")
  add(new JLabel(" "), "wrap")

  val saveButton = new JButton(t("button.save_deck_slots"))
  saveButton.addActionListener(ActionListener(_ => onEDT(saveDeckSlots())))
  add(saveButton, "skip")

  val refreshButton = new JButton(t("button.refresh"))
  refreshButton.addActionListener(ActionListener(_ =>
    try {
      onEDT(updateDecks())
    } catch {
      case e1: IOException => Main.showErrorDialog("Error updating decks", e1)
    }))
  add(refreshButton, "wrap")

  add(new JLabel(" "), "wrap")
  add(new JLabel(" "), "wrap")
  val exportButton = new JButton(t("button.export_deck"))
  exportButton.addActionListener(ActionListener(_ => onEDT(ExportDeckBox.open(monitor))))
  add(exportButton, "skip")

  add(new JLabel(" "), "wrap")
  val myDecksButton = new JButton(t("manage_decks_on_hsnet"))
  myDecksButton.addActionListener(ActionListener(_ => Browse(DECKS_URL)))
  add(myDecksButton, "skip,span")

  onEDT(updateDecks())

  def updateDecks() {
    DeckUtils.updateDecks()
    for (d <- deckSlotComboBoxes) d.applyDecks()
  }

  private def name(o: JSONObject): String = {
    Constants.hsClassOptions(Integer.parseInt(o.get("klass_id").toString)) +
      " - " +
      o.get("name").toString.toLowerCase
  }

  private def saveDeckSlots() {
    try {
      val slots = deckSlotComboBoxes map (_.selectedDeckId)
      API.setDeckSlots(slots)
      Main.showMessageDialog(this, API.message)
      updateDecks()
    } catch {
      case e: Exception => Main.showErrorDialog("Error saving deck slots", e)
    }
  }

  private def removeSlot(slot: Int): Unit = {
    if (JOptionPane.YES_OPTION == JOptionPane.showConfirmDialog(this, "Are you sure ?", "Remove a deck", JOptionPane.YES_NO_OPTION)) {
      for (i <- slot to 8) {
        val nextIndex = deckSlotComboBoxes(i).comboBox.getSelectedIndex
        deckSlotComboBoxes(i - 1).comboBox.setSelectedIndex(nextIndex)
      }
      deckSlotComboBoxes(8).comboBox.setSelectedIndex(0)
    }
  }

  private def createDeck(in: Option[Deck]): Unit = {
    in match {
      case Some(d) => {
        if (!d.isValid) {
          JOptionPane.showConfirmDialog(this,
            s"""${d.name} is not valid (${d.cardCount} cards). Do you want to edit it first on Hearthstats.net ?""".stripMargin) match {
              case JOptionPane.YES_OPTION =>
                Browse(s"http://hearthstats.net/decks/${d.slug}/edit")
              case JOptionPane.NO_OPTION => doCreate(d)
              case _ =>
            }
        } else doCreate(d)
      }
      case None => {
        JOptionPane.showMessageDialog(this,
          s"""There is no deck in this slot.
             |Choose a deck before pressing Construct""".stripMargin)
      }
    }

    def doCreate(d: Deck) = {
      monitor._hsHelper.bringWindowToForeground
      HsRobot(monitor._hsHelper.getHSWindowBounds).create(d)
    }
  }


  class DeckSlotPanel(slot: Int) extends JPanel {
    setLayout(new BorderLayout)
    add(new JLabel(t("deck_slot.label", slot)), BorderLayout.NORTH)
    val comboBox = new JComboBox[Object]
    add(comboBox, BorderLayout.CENTER)

    val removeBtn = new JButton("x") //TODO use an icon instead
    removeBtn.setToolTipText("Remove this deck and shift next ones")
    removeBtn.addActionListener(ActionListener(_ => removeSlot(slot)))
    add(removeBtn, BorderLayout.EAST)

    val createBtn = new JButton("Construct")
    createBtn.setToolTipText("""<html><b>Automatically creates this deck in Hearthstone</b><br/>
						        (providing you have the required cards)<br/>
						        <br/>
						        <i>You need to be in the collection mode and select the hero yourself</i>""")
    createBtn.addActionListener(ActionListener(_ => createDeck(selectedDeck)))

    add(createBtn, BorderLayout.SOUTH)

    def selectedDeck: Option[Deck] =
      comboBox.getSelectedItem match {
        case deck: Deck => Some(deck)
        case _ => None
      }

    def selectedDeckId: Option[Int] =
      selectedDeck.map(_.id)

    def applyDecks(): Unit = {
      comboBox.removeAllItems()
      comboBox.addItem(t("deck_slot.empty"))
      val decks = DeckUtils.getDeckLists.sortBy(d => (d.hero, d.name))
      for (deck <- decks) {
        comboBox.addItem(deck)
        if (deck.activeSlot == Some(slot)) comboBox.setSelectedItem(deck)
      }
    }
  }
}

File: uploader/src/main/java/net/hearthstats/ui/DecksTab.java
package net.hearthstats.ui;

import static net.hearthstats.util.Translations.t;

import java.awt.Desktop;
import java.awt.Dimension;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.IOException;
import java.net.URI;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import javax.swing.JButton;
import javax.swing.JComboBox;
import javax.swing.JLabel;
import javax.swing.JPanel;

import net.hearthstats.API;
import net.hearthstats.Constants;
import net.hearthstats.DeckUtils;
import net.hearthstats.Main;
import net.miginfocom.swing.MigLayout;

import org.json.simple.JSONObject;

public class DecksTab extends JPanel {
	private JComboBox<String> _deckSlot1Field;
	private JComboBox<String> _deckSlot2Field;
	private JComboBox<String> _deckSlot3Field;
	private JComboBox<String> _deckSlot4Field;
	private JComboBox<String> _deckSlot5Field;
	private JComboBox<String> _deckSlot6Field;
	private JComboBox<String> _deckSlot7Field;
	private JComboBox<String> _deckSlot8Field;
	private JComboBox<String> _deckSlot9Field;
	protected API _api = new API();
	private static final String DECKS_URL = "http://hearthstats.net/decks";

	public DecksTab() {

		MigLayout layout = new MigLayout();
		setLayout(layout);

		add(new JLabel(" "), "wrap");
		add(new JLabel(t("set_your_deck_slots")), "skip,span,wrap");
		add(new JLabel(" "), "wrap");

		add(new JLabel(t("deck_slot.label_1")), "skip");
		add(new JLabel(t("deck_slot.label_2")), "");
		add(new JLabel(t("deck_slot.label_3")), "wrap");

		_deckSlot1Field = new JComboBox<>();
		add(_deckSlot1Field, "skip");
		_deckSlot2Field = new JComboBox<>();
		add(_deckSlot2Field, "");
		_deckSlot3Field = new JComboBox<>();
		add(_deckSlot3Field, "wrap");

		add(new JLabel(" "), "wrap");

		add(new JLabel(t("deck_slot.label_4")), "skip");
		add(new JLabel(t("deck_slot.label_5")), "");
		add(new JLabel(t("deck_slot.label_6")), "wrap");

		_deckSlot4Field = new JComboBox<>();
		add(_deckSlot4Field, "skip");
		_deckSlot5Field = new JComboBox<>();
		add(_deckSlot5Field, "");
		_deckSlot6Field = new JComboBox<>();
		add(_deckSlot6Field, "wrap");

		add(new JLabel(" "), "wrap");

		add(new JLabel(t("deck_slot.label_7")), "skip");
		add(new JLabel(t("deck_slot.label_8")), "");
		add(new JLabel(t("deck_slot.label_9")), "wrap");

		_deckSlot7Field = new JComboBox<>();
		add(_deckSlot7Field, "skip");
		_deckSlot8Field = new JComboBox<>();
		add(_deckSlot8Field, "");
		_deckSlot9Field = new JComboBox<>();
		add(_deckSlot9Field, "wrap");

		add(new JLabel(" "), "wrap");
		add(new JLabel(" "), "wrap");

		JButton saveButton = new JButton(t("button.save_deck_slots"));
		saveButton.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				_saveDeckSlots();
			}
		});
		add(saveButton, "skip");

		JButton refreshButton = new JButton(t("button.refresh"));
		refreshButton.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				try {
					updateDecks();
				} catch (IOException e1) {
					Main.showErrorDialog("Error updating decks", e1);
				}
			}
		});
		add(refreshButton, "wrap,span");

		add(new JLabel(" "), "wrap");
		add(new JLabel(" "), "wrap");

		JButton myDecksButton = new JButton(t("manage_decks_on_hsnet"));
		myDecksButton.addActionListener(new ActionListener() {
			@Override
			public void actionPerformed(ActionEvent e) {
				try {
					Desktop.getDesktop().browse(new URI(DECKS_URL));
				} catch (Throwable e1) {
					Main.showErrorDialog("Error launching browser with URL"
							+ DECKS_URL, e1);
				}
			}
		});
		add(myDecksButton, "skip,span");

	}

	public void updateDecks() throws IOException {
		DeckUtils.updateDecks();
		_applyDecksToSelector(_deckSlot1Field, 1);
		_applyDecksToSelector(_deckSlot2Field, 2);
		_applyDecksToSelector(_deckSlot3Field, 3);
		_applyDecksToSelector(_deckSlot4Field, 4);
		_applyDecksToSelector(_deckSlot5Field, 5);
		_applyDecksToSelector(_deckSlot6Field, 6);
		_applyDecksToSelector(_deckSlot7Field, 7);
		_applyDecksToSelector(_deckSlot8Field, 8);
		_applyDecksToSelector(_deckSlot9Field, 9);
	}

	private void _applyDecksToSelector(JComboBox<String> selector,
			Integer slotNum) {

		selector.setMaximumSize(new Dimension(200, selector.getSize().height));
		selector.removeAllItems();

		selector.addItem("- Select a deck -");

		List<JSONObject> decks = DeckUtils.getDecks();

		Collections.sort(decks, new Comparator<JSONObject>() {
			@Override
			public int compare(JSONObject o1, JSONObject o2) {
				return name(o1).compareTo(name(o2));
			}

		});

		for (int i = 0; i < decks.size(); i++) {
			selector.addItem(name(decks.get(i))
					+ "                                       #"
					+ decks.get(i).get("id"));
			if (decks.get(i).get("slot") != null
					&& decks.get(i).get("slot").toString()
							.equals(slotNum.toString()))
				selector.setSelectedIndex(i + 1);
		}
	}

	private String name(JSONObject o) {
		return Constants.hsClassOptions[Integer.parseInt(o.get("klass_id")
				.toString())]
				+ " - " + o.get("name").toString().toLowerCase();
	}

	private void _saveDeckSlots() {

		try {
			_api.setDeckSlots(_getDeckSlotDeckId(_deckSlot1Field),
					_getDeckSlotDeckId(_deckSlot2Field),
					_getDeckSlotDeckId(_deckSlot3Field),
					_getDeckSlotDeckId(_deckSlot4Field),
					_getDeckSlotDeckId(_deckSlot5Field),
					_getDeckSlotDeckId(_deckSlot6Field),
					_getDeckSlotDeckId(_deckSlot7Field),
					_getDeckSlotDeckId(_deckSlot8Field),
					_getDeckSlotDeckId(_deckSlot9Field));
			Main.showMessageDialog(this, _api.getMessage());
			updateDecks();
		} catch (Throwable e) {
			Main.showErrorDialog("Error saving deck slots", e);
		}
	}

	private Integer _getDeckSlotDeckId(JComboBox selector) {
		Integer deckId = null;
		String deckStr = (String) selector.getItemAt(selector
				.getSelectedIndex());
		Pattern pattern = Pattern.compile("[^0-9]+([0-9]+)$");
		Matcher matcher = pattern.matcher(deckStr);
		if (matcher.find()) {
			deckId = Integer.parseInt(matcher.group(1));
		}
		return deckId;
	}
}
